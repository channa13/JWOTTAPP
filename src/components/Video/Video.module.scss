@use '../../styles/variables';
@use '../../styles/theme';
@use '../../styles/mixins/responsive';

.video {
  color: var(--primary-color);
  font-family: var(--body-font-family);
  font-weight: 400;
  text-shadow: 0 2px 4px rgba(0, 0, 0, 0.14), 0 3px 4px rgba(0, 0, 0, 0.12), 0 1px 5px rgba(0, 0, 0, 0.2);
}

.mainPadding {
  padding: 37px 56px 0;

  @include responsive.desktop-small-only() {
    padding: 34px 36px 36px;
  }

  @include responsive.tablet-only() {
    padding: 34px 24px 24px;
  }

  @include responsive.mobile-only() {
    padding: 16px;
  }
}

.main {
  display: flex;
  max-width: 100vw;
  overflow: hidden;

  &.hidden {
    display: none;
  }

  &.posterNormal {
    position: relative;
  }
}

.info {
  width: 70%;
  max-width: 650px;

  @include responsive.tablet-only() {
    width: 350px;
    min-height: 380px;
  }

  @include responsive.mobile-only() {
    width: 100%;
    min-height: 380px;
    padding-top: 225px;
  }
}

.title {
  color: var(--primary-color);
  font-family: var(--body-font-family);
  font-weight: var(--body-font-weight-bold);
  font-size: 34px;
  line-height: 40px;
  letter-spacing: 0.25px;

  text-shadow: 0 2px 4px rgba(0, 0, 0, 0.14), 0 3px 4px rgba(0, 0, 0, 0.12), 0 1px 5px rgba(0, 0, 0, 0.2);

  @include responsive.mobile-only() {
    font-size: 20px;
    line-height: 20px;
    letter-spacing: 0.15px;
  }
}

.metaContainer {
  display: flex;
  flex-direction: column;
}

.meta {
  margin-bottom: 8px;
  font-size: 16px;
  line-height: 18px;
  letter-spacing: 0.15px;

  @include responsive.mobile-only() {
    order: 2;
    font-size: 14px;
  }
}

.seriesMeta {
  margin-top: 24px;
  margin-bottom: 8px;
  font-size: 20px;
  line-height: 20px;
  letter-spacing: 0.5px;

  &.seriesMetaPlayer {
    margin-top: 4px;
  }

  @include responsive.mobile-only() {
    order: 1;
    margin: 4px 0;
    font-size: 18px;
  }
}

.description {
  min-height: 21px;
  margin-bottom: 24px;
  font-size: 18px;
  line-height: 20px;
  letter-spacing: 0.5px;
}

.buttonBar {
  display: flex;
  justify-content: flex-start;
  align-items: center;
  width: 100%;
  margin: 0 -6px;

  > button {
    justify-content: center;
    margin: 6px;
  }

  @include responsive.mobile-and-tablet() {
    flex-wrap: wrap;

    > button {
      flex: 1;
      padding: 0 12px;
      
      &:first-child {
        flex-basis: 100%;
      }
    }
  }

  @include responsive.mobile-only() {
    flex-wrap: wrap;

    > button:first-child {
      margin-bottom: 8px;
    }
  }
}

.poster {
  position: absolute;
  top: 0;
  right: 0;
  z-index: -1;
  background-repeat: no-repeat;
  background-position: center center;
  background-size: cover;

  &.normal {
    width: 720px;
    height: 405px;
    margin: 37px 56px;
    border-radius: 4px;
    mask-image: linear-gradient(-90deg, rgba(0, 0, 0, 1) 60%, rgba(0, 0, 0, 0) 100%);
    -webkit-mask-image: linear-gradient(-90deg, rgba(0, 0, 0, 1) 60%, rgba(0, 0, 0, 0) 100%); /* stylelint-disable-line */

    @include responsive.tablet-only() {
      width: 600px;
      height: 338px;
      margin: 24px 34px;
    }

    @include responsive.mobile-only() {
      width: 100vw;
      margin: 0 16px;

      mask-image: radial-gradient(farthest-corner at 80% 20%, rgba(0, 0, 0, 1) 40%, rgba(0, 0, 0, 0) 60%);
      //noinspection CssInvalidPropertyValue
      -webkit-mask-image: radial-gradient(farthest-corner at 80% 20%, rgba(0, 0, 0, 1) 40%, rgba(0, 0, 0, 0) 60%); /* stylelint-disable-line */
    }
  }

  &.fading {
    width: 80vw;
    height: calc(85vw / 16 * 9);
    mask-image: radial-gradient(farthest-corner at 80% 20%, rgba(0, 0, 0, 1) 30%, rgba(0, 0, 0, 0) 60%);
    //noinspection CssInvalidPropertyValue
    -webkit-mask-image: radial-gradient(farthest-corner at 80% 20%, rgba(0, 0, 0, 1) 30%, rgba(0, 0, 0, 0) 60%); /* stylelint-disable-line */

    @include responsive.tablet-only() {
      width: 80vw;
      height: calc(140vw / 16 * 9);
    }

    @include responsive.mobile-only() {
      width: 110vw;
      height: calc(180vw / 16 * 9);
    }
  }
}

.playIcon {
  position: absolute;
  top: calc(50% - 60px);
  left: calc(50% - 35px);
  padding: 25px 30px;
  font-size: xx-large;
  border: 3px solid white;
  border-radius: 100%;
}

.related {
  position: relative;
}

.playerContainer {
  position: fixed;
  top: 0;
  left: 0;
  z-index: variables.$player-z-index;
  width: 100vw;
  height: calc(100vh - calc(100vh - 100%));
  overflow: hidden;
  background-color: variables.$black;
}

.player {
  position: absolute;
  top: 0;
  left: 0;
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  height: 100%;
}

.playerOverlay {
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  min-height: 100px;
  padding: 24px 56px;
  background-image: linear-gradient(to top, rgba(0, 0, 0, 0) 0%, rgba(0, 0, 0, 0.1) 25%, rgba(0, 0, 0, 0.25) 100%);
  pointer-events: none;

  @include responsive.mobile-only() {
    padding-left: 16px;
  }
}

.playerContent {
  display: flex;
  width: 80%;

  @include responsive.mobile-only() {
    width: 100%;
  }
}

.backButton {
  margin-right: 24px;
  opacity: 1;
  pointer-events: auto;

  > svg {
    width: 36px;
    height: 36px;
    -webkit-filter: drop-shadow(2px 2px 5px rgba(0, 0, 0, 0.4)); /* stylelint-disable-line */
    filter: drop-shadow(2px 2px 5px rgba(0, 0, 0, 0.4));
  }
}

.trailerModal {
  position: relative;
  width: 80vw;
  height: calc(80vw / 16 * 9);

  @include responsive.mobile-only() {
    width: 100vw;
    height: calc(100vw / 16 * 9);
  }
}

.trailerMeta {
  position: absolute;
  top: 16px;
  left: 16px;
  z-index: 1;

  &.hidden {
    opacity: 0;
    transition: opacity 0.6s ease;
  }
}
