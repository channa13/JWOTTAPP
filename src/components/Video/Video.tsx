import React, { useCallback, useEffect, useState } from 'react';
import classNames from 'classnames';
import { useTranslation } from 'react-i18next';

import styles from './Video.module.scss';

import CollapsibleText from '#src/components/CollapsibleText/CollapsibleText';
import Button from '#src/components/Button/Button';
import IconButton from '#src/components/IconButton/IconButton';
import Modal from '#src/components/Modal/Modal';
import Fade from '#src/components/Animation/Fade/Fade';
import Alert from '#src/components/Alert/Alert';
import ModalCloseButton from '#src/components/ModalCloseButton/ModalCloseButton';
import Cinema from '#src/containers/Cinema/Cinema';
import useBreakpoint, { Breakpoint } from '#src/hooks/useBreakpoint';
import Favorite from '#src/icons/Favorite';
import PlayTrailer from '#src/icons/PlayTrailer';
import Share from '#src/icons/Share';
import Check from '#src/icons/Check';
import ArrowLeft from '#src/icons/ArrowLeft';
import { formatDuration } from '#src/utils/formatting';
import FavoriteBorder from '#src/icons/FavoriteBorder';
import type { PlaylistItem } from '#types/playlist';
import { useFavoritesStore } from '#src/stores/FavoritesStore';

type Poster = 'fading' | 'normal';

type Props = {
  title: string;
  item: PlaylistItem;
  feedId?: string;
  trailerItem?: PlaylistItem;
  play: boolean;
  goBack: () => void;
  onComplete?: () => void;
  isFavorited: boolean;
  isFavoritesEnabled: boolean;
  onFavoriteButtonClick: () => void;
  poster: Poster;
  enableSharing: boolean;
  hasShared: boolean;
  onShareClick: () => void;
  playTrailer: boolean;
  onTrailerClick: () => void;
  onTrailerClose: () => void;
  isSeries?: boolean;
  episodeCount?: number;
  startWatchingButton: JSX.Element;
  children?: JSX.Element;
};

const Video: React.FC<Props> = ({
  title,
  item,
  feedId,
  trailerItem,
  play,
  goBack,
  onComplete,
  poster,
  enableSharing,
  hasShared,
  onShareClick,
  isFavorited,
  isFavoritesEnabled,
  onFavoriteButtonClick,
  children,
  playTrailer,
  onTrailerClick,
  onTrailerClose,
  isSeries = false,
  startWatchingButton,
  episodeCount,
}: Props) => {
  const [isPlaying, setIsPlaying] = useState<boolean>(false);
  const [userActive, setUserActive] = useState(true);
  const breakpoint: Breakpoint = useBreakpoint();
  const { t } = useTranslation(['video', 'common']);

  const handleUserActive = useCallback(() => setUserActive(true), []);
  const handleUserInactive = useCallback(() => setUserActive(false), []);
  const handlePlay = useCallback(() => setIsPlaying(true), []);
  const handlePause = useCallback(() => setIsPlaying(false), []);
  const handleComplete = useCallback(() => onComplete && onComplete(), [onComplete]);

  const { clearWarning, warning } = useFavoritesStore((state) => ({
    clearWarning: state.clearWarning,
    warning: state.warning,
  }));

  const isLargeScreen = breakpoint >= Breakpoint.md;
  const isMobile = breakpoint === Breakpoint.xs;
  const imageSourceWidth = 640 * (window.devicePixelRatio > 1 || isLargeScreen ? 2 : 1);
  const posterImage = item.image.replace('720', imageSourceWidth.toString()); // Todo: should be taken from images (1280 should be sent from API)

  const metaData = [];
  if (item.pubdate) metaData.push(new Date(item.pubdate * 1000).getFullYear());
  if (!isSeries && item.duration) metaData.push(formatDuration(item.duration));
  if (isSeries && episodeCount) metaData.push(t('video:total_episodes', { count: episodeCount }));
  if (item.genre) metaData.push(item.genre);
  if (item.rating) metaData.push(item.rating);
  const metaString = metaData.join(' â€¢ ');

  const seriesMeta = isSeries && (
    <>
      <strong>{`S${item.seasonNumber}:E${item.episodeNumber}`}</strong>
      {' - '}
      {item.title}
    </>
  );

  useEffect(() => {
    if (play || playTrailer) setUserActive(true);
  }, [play, playTrailer]);

  return (
    <div className={styles.video}>
      <div
        className={classNames(styles.main, styles.mainPadding, {
          [styles.posterNormal]: poster === 'normal',
        })}
      >
        <div className={styles.info}>
          <h2 className={styles.title}>{title}</h2>
          <div className={styles.metaContainer}>
            <div className={styles.meta}>{metaString}</div>
            {isSeries && <div className={styles.seriesMeta}>{seriesMeta}</div>}
          </div>
          <CollapsibleText text={item.description} className={styles.description} maxHeight={isMobile ? 60 : 'none'} />

          <div className={styles.buttonBar}>
            {startWatchingButton}
            {trailerItem && (
              <Button
                className={styles.bigButton}
                label={t('video:trailer')}
                aria-label={t('video:watch_trailer')}
                startIcon={<PlayTrailer />}
                onClick={onTrailerClick}
                active={playTrailer}
                fullWidth={breakpoint < Breakpoint.md}
              />
            )}
            {isFavoritesEnabled && (
              <Button
                label={t('video:favorite')}
                aria-label={isFavorited ? t('video:remove_from_favorites') : t('video:add_to_favorites')}
                startIcon={isFavorited ? <Favorite /> : <FavoriteBorder />}
                onClick={onFavoriteButtonClick}
                color={isFavorited ? 'primary' : 'default'}
                fullWidth={breakpoint < Breakpoint.md}
              />
            )}
            {enableSharing && (
              <Button
                label={hasShared ? t('video:copied_url') : t('video:share')}
                startIcon={hasShared ? <Check /> : <Share />}
                onClick={onShareClick}
                active={hasShared}
                fullWidth={breakpoint < Breakpoint.md}
              />
            )}
          </div>
        </div>
        <div className={classNames(styles.poster, styles[poster])} style={{ backgroundImage: `url('${posterImage}')` }} />
      </div>
      {!!children && <div className={classNames(styles.related, styles.mainPadding)}>{children}</div>}
      <Fade open={play}>
        <div className={styles.playerContainer}>
          <div className={styles.player}>
            <Cinema
              item={item}
              feedId={feedId}
              onPlay={handlePlay}
              onPause={handlePause}
              onComplete={handleComplete}
              onUserActive={handleUserActive}
              onUserInActive={handleUserInactive}
            />
          </div>
          <Fade open={!isPlaying || userActive}>
            <div className={styles.playerOverlay}>
              <div className={styles.playerContent}>
                <IconButton aria-label={t('common:back')} onClick={goBack} className={styles.backButton}>
                  <ArrowLeft />
                </IconButton>
                <div>
                  <h2 className={styles.title}>{title}</h2>
                  <div className={styles.metaContainer}>
                    {isSeries && <div className={classNames(styles.seriesMeta, styles.seriesMetaPlayer)}>{seriesMeta}</div>}
                    <div className={styles.meta}>{metaString}</div>
                  </div>
                </div>
              </div>
            </div>
          </Fade>
        </div>
      </Fade>
      <Alert open={warning !== null} message={warning} onClose={clearWarning} />
      {!!trailerItem && (
        <Modal open={playTrailer} onClose={onTrailerClose}>
          <div className={styles.trailerModal}>
            <Cinema
              item={trailerItem}
              onPlay={handlePlay}
              onPause={handlePause}
              onComplete={onTrailerClose}
              onUserActive={handleUserActive}
              onUserInActive={handleUserInactive}
              isTrailer
            />
            <div className={classNames(styles.playerOverlay, { [styles.hidden]: isPlaying && !userActive })} />
            <div className={classNames(styles.trailerMeta, styles.title, { [styles.hidden]: isPlaying && !userActive })}>{`${title} - Trailer`}</div>
            <ModalCloseButton onClick={onTrailerClose} visible={!isPlaying || userActive} />
          </div>
        </Modal>
      )}
    </div>
  );
};

export default Video;
