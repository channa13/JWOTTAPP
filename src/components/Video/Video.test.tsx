import React from 'react';
import { render } from '@testing-library/react';

import Video from './Video';

import type { PlaylistItem } from '#types/playlist';

describe('<Video>', () => {
  test('renders and matches snapshot', () => {
    const item = {
      description: 'Test item description',
      duration: 354,
      feedid: 'ax85aa',
      image: 'http://test/img.jpg',
      images: [],
      link: 'http://test/link',
      genre: 'Tester',
      mediaid: 'zp50pz',
      pubdate: 26092021,
      rating: 'CC_CC',
      sources: [],
      seriesId: 'ag94ag',
      tags: 'Test tag',
      title: 'Test item title',
      tracks: [],
    } as PlaylistItem;
    const { container } = render(
      <Video
        title="Test video"
        play
        item={item}
        goBack={vi.fn()}
        poster="fading"
        hasShared={false}
        onShareClick={vi.fn()}
        enableSharing
        isFavorited={false}
        isFavoritesEnabled={true}
        onFavoriteButtonClick={vi.fn()}
        playTrailer={false}
        onTrailerClick={vi.fn()}
        onTrailerClose={vi.fn()}
        startWatchingButton={<button>Start watching</button>}
      />,
    );

    expect(container).toMatchSnapshot();
  });
});
