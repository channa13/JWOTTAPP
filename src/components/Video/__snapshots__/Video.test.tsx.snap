// Vitest Snapshot v1

exports[`<Video> > renders and matches snapshot 1`] = `
<div>
  <div
    class="video"
  >
    <div
      class="main mainPadding"
    >
      <div
        class="info"
      >
        <h2
          class="title"
        >
          Test video
        </h2>
        <div
          class="metaContainer"
        >
          <div
            class="meta"
          >
            1970 • 6m  • Tester • CC_CC
          </div>
        </div>
        <div
          class="collapsibleText"
        >
          <div
            class="textContainer description"
            style="max-height: none;"
          >
            Test item description
          </div>
        </div>
        <div
          class="buttonBar"
        >
          <button>
            Start watching
          </button>
          <button
            aria-label="video:add_to_favorites"
            class="button default outlined"
          >
            <div
              class="startIcon"
            >
              <svg
                aria-hidden="true"
                class="icon"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <path
                    d="M12.1,18.55L12,18.65L11.89,18.55C7.14,14.24 4,11.39 4,8.5C4,6.5 5.5,5 7.5,5C9.04,5 10.54,6 11.07,7.36H12.93C13.46,6 14.96,5 16.5,5C18.5,5 20,6.5 20,8.5C20,11.39 16.86,14.24 12.1,18.55M16.5,3C14.76,3 13.09,3.81 12,5.08C10.91,3.81 9.24,3 7.5,3C4.42,3 2,5.41 2,8.5C2,12.27 5.4,15.36 10.55,20.03L12,21.35L13.45,20.03C18.6,15.36 22,12.27 22,8.5C22,5.41 19.58,3 16.5,3Z"
                  />
                </g>
              </svg>
            </div>
            <span>
              video:favorite
            </span>
          </button>
          <button
            class="button default outlined"
          >
            <div
              class="startIcon"
            >
              <svg
                aria-hidden="true"
                class="icon"
                focusable="false"
                viewBox="0 0 24 24"
                xmlns="http://www.w3.org/2000/svg"
              >
                <g>
                  <path
                    d="M18,16.08C17.24,16.08 16.56,16.38 16.04,16.85L8.91,12.7C8.96,12.47 9,12.24 9,12C9,11.76 8.96,11.53 8.91,11.3L15.96,7.19C16.5,7.69 17.21,8 18,8A3,3 0 0,0 21,5A3,3 0 0,0 18,2A3,3 0 0,0 15,5C15,5.24 15.04,5.47 15.09,5.7L8.04,9.81C7.5,9.31 6.79,9 6,9A3,3 0 0,0 3,12A3,3 0 0,0 6,15C6.79,15 7.5,14.69 8.04,14.19L15.16,18.34C15.11,18.55 15.08,18.77 15.08,19C15.08,20.61 16.39,21.91 18,21.91C19.61,21.91 20.92,20.61 20.92,19A2.92,2.92 0 0,0 18,16.08Z"
                  />
                </g>
              </svg>
            </div>
            <span>
              video:share
            </span>
          </button>
        </div>
      </div>
      <div
        class="poster fading"
        style="background-image: url(http://test/img.jpg);"
      />
    </div>
    <div
      style="transition: opacity 0.25s ease-in-out; opacity: 0;"
    >
      <div
        class="playerContainer"
      >
        <div
          class="player"
        >
          <div
            class="cinema fill"
          >
            <div />
          </div>
        </div>
        <div
          style="transition: opacity 0.25s ease-in-out; opacity: 0;"
        >
          <div
            class="playerOverlay"
          >
            <div
              class="playerContent"
            >
              <div
                aria-label="common:back"
                class="iconButton backButton"
                role="button"
                tabindex="0"
              >
                <svg
                  aria-hidden="true"
                  class="icon"
                  focusable="false"
                  viewBox="0 0 24 24"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <g>
                    <path
                      d="M20,11V13H8L13.5,18.5L12.08,19.92L4.16,12L12.08,4.08L13.5,5.5L8,11H20Z"
                    />
                  </g>
                </svg>
              </div>
              <div>
                <h2
                  class="title"
                >
                  Test video
                </h2>
                <div
                  class="metaContainer"
                >
                  <div
                    class="meta"
                  >
                    1970 • 6m  • Tester • CC_CC
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
